### ISSUE TITLE
Modules

### ISSUE THREAD

**Issue by cardillan** [cardillan/mindcode#149]
A draft for adding support for modules to Mindcode. To be implemented in 2 phases.

NOTE: _parameters_ in this text aren't function parameters, but global variables declared using the `param` keyword and used to allow parametrization of the compiled mlog code. 

**1st phase - Support for loading additional files**

* `require` keyword for loading other source files. Takes a filename as an argument (`require "../library/system.mnd"`).
* Each loaded file would be parsed separately. After parsing the topmost file and all loaded files, the parse trees would be combined.
    * Loads would be handled recursively. File which was already processed wouldn't be processed again --> cycles in module dependencies are allowed.
        * Paths are resolved relative to the file containing the `require` declaration (is this optimal?) 
        * File paths would be converted to canonical paths to identify already processed files.
        * Given the inherently limited size od Mindcode projects, I don't expect complicated dependencies between modules.
    * Syntax errors will be reported per file.
* The entrypoint (`begin ... end` or alternatively `def main()`), if present, must be contained in the main file.
* Rules for combining parse trees:
    * Parse trees are combined at the topmost level only (topmost nodes merged together).
        * AST contexts assigned to the nodes will contain information about the source file and module 
    * `allocate stack` and `allocate heap`, if declared multiple times, must be declared identically.
    * A function must be declared only once.
    * A constant or parameter can be declared multiple times if the assigned value is always the same.
        * Perhaps a constant or parameter in a module might be declared without initial value, which would be provided by the main module.
        * Alternatively, value assigned by a module can be reassigned by the main module.
    * A global array/variable can be declared multiple times either with initial value or without it. The initial values, if assigned, must be the same.
    * The combined parse tree will be compiled/optimized.
        * Optimizations run once against the entire compiled code.
        * Everything will be properly resolved, no compilation against uncertain/unknown constructs.

**2nd phase - Modules, namespaces and a system library**

* **Modules**
    * Module names specified using `module` keyword: `module foo;`
        * A Java/C# convention for module names - identifiers separated by a dot: `module cardillan.drawing.icons;`
    * `module` needs to be the first declaration in the file.
    * The first loaded file contains the main module. Main module has no name (no `module` declaration). Main module cannot be required.
    * Each module has its own namespace named after the module identifier. The main module uses global namespace.
        * A command to import default objects from a given module into the global namespace
            * `import foo.bar;` - imports just one identifier from `foo`
            * `import foo.*;` - imports all
    * Visibility: public/private/default - needs a lot of thought!
        * A public object is always included in the global namespace (parameters are always public)
            * Probably not allowed for functions - avoid global namespace pollution. 
        * A private object is inaccessible outside the module (preventing the access is additional work, possibly in a 3rd phase)
        * A default object can be accessed via full module name or imported into the global namespace using `import`  
    * No support for splitting modules into several files until really needed, hopefully never
        * The same module declared in two different loaded files causes an error
* **Compilation of modules without a main module**
    * A module can be compiled separately.
    * If the module doesn't contain an entrypoint, all functions would be considered reachable and wouldn't be removed by unreachable code elimination.
    * If all dependencies of the module are resolved, a valid, optimized mlog code would be produced
        * Easy verification the module actually compiles
        * Generated code may be inspected
* **Target Mindcode versions**
    * Modules may indicate a range of Mindustry Logic versions they support.
        * `#target ver;` or `#target ver1 .. ver2;` (trying not to allocate new keywords for this).
        * The way of labeling the Mindustry Logic versions has not been specified yet, maybe using build numbers.
        * When not specified, all versions must be supported.
    * Loading a module incompatible with the current target results into an error
* **System modules**
    * System modules will contain a system library of functions.
    * System modules will reside in GitHub Mindcode repository and **will be implicitly accessible, even from the web app**
        * Compiled on each build to make sure they work, using the latest supported Mindustry Logic version specified by the module as the target
        * System modules would be required by module names, not by file names.
            * `require standard;` or `require cardillan.drawing.icons;` - no double quotes
            * Source files for these modules would be located and imported automatically based on module name, using a config file automatically generated during the test build assigning source files to module names.
            * Parse trees of system modules generated during the test build could be stored to save time - would make sense if we build a really, really large collection of system functions.
                * Unused functions from such a large parse tree wouldn't be even compiled - no time wasted on them.       
        * Contributions to system modules would need to be carefully vetted
        * Module `standard`: system module for standard library functions (e.g. `findUnit`)
        * Module `wp` to enclose world processor instructions.
            * `import wp.*;` would be required, unless we make an exception and allow public functions in module `wp` (or in all system modules).  
            * No more standard/world targets. 
    * Host other modules ("user modules"?) in the repository as contributions from different users
        * Included via pull requests
            * No vetting process, except that they compile
            * If they fail to compile, they'll be removed from the release - probably via some config file listing defunct user modules
        * May need establishing a naming convention for user modules, perhaps after github usernames (e.g. `cardillan.buildings`, `cardillan.drawings` etc.)
        * Instead of `require cardillan.drawing.icons` use `require "cardillan/drawing/icons.mnd"` to develop against a local version of a module.

Example:

```
require standard;

begin
    unit = standard.findUnit(@mega);
    // Do something with unit
end;
```

Importing two global variables might merge them - do we want to allow this? Might be useful in some situations - e.g. if two modules want to share variables referencing to units.

```
module foo;
var debug;
...

module bar;
var debug;          // Variables not public - this is different from foo.debug!

require foo, bar;
import foo.debug
import bar.debug;   // Merges foo.debug with bar.debug. Or error. 
```

**Rejected ideas**

* Replace language targets with modules: the program would include the module for the proper target
    * What if two modules include conflicting target modules?
* Support for different targets in one module: have sections of code only be included when compiling for given target. That way the same logic could be implemented using instruction sets from different targets.
    * Support for different Mindustry versions doesn't seem to be crucial.
    * Can be still made by moving target-dependent logic into separate modules and including them as necessary.
* Using C-like `#include` directive.
    * The included file can be put anywhere - say, inside a loop. That means it generally cannot be parsed separately.
    * There's no direct support from ANTLR to do this, so it means either using/writing a preprocessor, or finding a correct way for ANTLR To do it. All seem to be a lot of work.
    * Line numbers would have to be fixed when reporting syntax errors.
* Compiled libraries
    * Each source would compile separately and the resulting code would be combined (linked) somehow
    * Lots and lots of work:
        * Needs a linker
        * Every file would be optimized separately - a global optimization step might be required after linking
    * Mindcode will never be used for such large projects that separate compilation of modules could save some significant time.

**Comment by cardillan** [comment]
The first phase has been more or less completed as of release 2.6.0. On to of that, system libraries were also implemented.

What remains to be done is namespaces and modules.

**Comment by cardillan** [comment]
### Current status

* Remote modules
  * When `require`d from another module, exported functions and variables become part of a namespace named after the remote processor.
  * When compiled separately, only the exported functions are considered entrypoints. No need to change.
  * Remote modules might need explicit declaration, i.e. `remote module name;` instead of `module name;`.
* Local modules:
  * When `require`d from another module/program, all functions and variables become part of the global namespace.
  * No support for private functions/variables.
  * No support for standalone compilation (compiled as remote modules, having no exported functions means no code gets produced).
* Namespaces
  * Remote modules reside in separate namespaces. The current implementation is satisfactory.
* Function/variable resolution:
  * A module may access functions and variables declared in a program (or theoretically in other modules). Such a module cannot be compiled standalone, but it currently provides functionality for modules to rely on external implementations. This needs to be resolved somehow (e.g., by supporting abstract functions). However, the resolution might also be postponed until classes are supported.

### Objectives

The module namespace is declared in the `module` statement. The final solution should allow namespace hierarchy there (e.g., `module cardillan.graphics.icons;`). When the file is included (e.g., by `require "cardillan/graphics/icons.mnd"` - note that Mindcode won't require the namespace hierarchy to match the directory hierarchy), the `cardillan.graphics.icons` namespace will become available and will contain all the (non-private) symbols from the included file.

However, having to specify `cardillan.graphics.icons.drawError()` for function calls is not a workable solution.

There are two different concerns that need to be addressed:

* adding a module/library to compilation,
* managing namespaces.

For example, in Java, the first concern is managed through project directory structure, and the second one is managed through the `import` statement. In Mindcode, the first point is managed through `require` (and I don't want to change that to an automatic source file discovery). I don't want to add a separate statement for importing module content into global namespace, as this would effectively result in two statements per one module. 

### Proposed solution

Everything declared in a program (i.e., a source file which doesn't contain a `module` declaration) becomes part of the global namespace. No public or private functions/variables are supported. A name declared in a program shadows any other name imported from a module into the global namespace; to access a shadowed name, a fully qualified name must be used.

For modules, the following rules will apply:

* A function or variable can be declared `public`, `private` or omit the scope qualifier.
* Functions/variables declared within a module are part of the global namespace for the module. Such a name shadows any other name imported from a module into the global namespace; to access a shadowed name, a fully qualified name must be used.
* `private` functions and variables are only available in given module and do not require namespace qualification.
* `public` functions and variables are automatically imported into the global namespace. This is meant basically for system libraries only.
* Functions and variables without scope qualifier are accessible via full qualification.
* The `require` statement will support a `namespace` clause. It will be possible to specify a namespace that will contain the non-private names from the module. Example: `require "cardillan/graphics/icons.mnd" namespace icons;` will allow calling `icons.drawError()`. (Needs a way to import into the global namespace explicitly, perhaps a `global` identifier/keyword for a global namespace?)
* Alternatively, omitting the `namespace` clause will import the module into the namespace named after the last part of the module namespace. Example: `require "cardillan/graphics/icons.mnd";` will allow calling `icons.drawError()`.
* Trying to import two different modules into the same namespace will result in a compilation error.
* The same module may end up being imported into different namespaces by different modules within the same program. The namespace imports therefore need to be local to the module.

In the future, we might have classes, but even with classes Mindcode will still be a heavily procedural language. I don't think the above principles should be different with classes.

**Comment by JeanJPNM** [comment]
For qualified import syntax, you could look at the way Dart [handles importing modules](https://dart.dev/language/libraries#specifying-a-library-prefix) since they also import all exported symbols into the importer's global scope by default. But they allow using `as name` to import under a namespace, `show foo, bar` to only import specific symbols and `hide foo, bar` to import everything except certain symbols.

**Comment by cardillan** [comment]
Thanks, that's precisely what I was looking for! Looks like the `import ... as` corresponds pretty much to my `require ... namespace`.

Another thing I want to implement is providing access to mlog instructions via namespaces. There should be a separate `world` (?) namespace for privileged instructions, and probably namespaces for individual instructions (e.g. `draw` namespace for the `draw` instruction, giving `draw.rectangle()`, `draw.print()` etc. - I believe this is what mlogjs does), because the current system has significant drawbacks. The old system would be available via some compatibility module.

There's one thing where I'm not entirely clear at this moment: the mlog namespaces might be provided implicitly by the compiler, or they might be stored in actual Mindcode source files (these files would be machine generated from instruction metadata). There's certain elegance in the second approach, but I'd need a way to store multiple namespaces in a single file, and also select different module for different targets (i.e., version 6/7/8.0/8.1 specific module). Both of these would require additional support in the syntax. Both of these could be useful for a future Mindcode system library.

There's a potential problem with parsing these files incurring a performance penalty. But firstly, current parser is reasonably fast (thanks to the syntax not being ambiguous), and if parsing time becomes a problem, I plan storing pre-parsed AST structures which would be generated during build, so these concerns are manageable. There are no caching mechanisms planned on derived structures (e.g., call graph and function definitions), but again, should this become a problem, it should be manageable somehow.

The instructions would be implemented by inline functions consisting of `mlog()`, `mlogText()` and `mlogSafe()` functions. Code generation for these calls would be probably a wee bit more complex/slower than the current specialized mlog mapper approach. Also, I'd probably need specific data types for keywords (the existing keyword categories are listed [here](https://github.com/cardillan/mindcode/blob/devel/doc/syntax/SYNTAX-EXTENSIONS.markdown#declaring-new-mlog-keywords)). So I might need to wait with this until data types get implemented (I plan to do this after namespaces).
